sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+        GValue *value = g_slice_new0(GValue);
+        g_value_init(value, G_TYPE_STRING);
+        g_value_set_string(value, user.toLocal8Bit());
+        lu_ent_add(ent, LU_MEMBERNAME, value);
+        if (!lu_group_modify(context, ent, &error)) {
+            qWarning() << "Group modify failed:" << lu_strerror(error);
+            rv = false;
+        }
+    } else {
+        qWarning() << "Could not find group";
+        rv = false;
+    }
+
+    lu_ent_free(ent);
+    lu_end(context);
+
+    return rv;
+}
+
+bool SailfishUserManager::addUserToGroups(const QString& user)
+{
+    QFile file(GROUPS_IDS_FILE);
+    if (!file.open(QIODevice::ReadOnly)) {
+        qWarning() << "Failed to open groups file";
+        return false;
+    }
+
+    QByteArray line;
+
+    while (!file.atEnd()) {
+        line = file.readLine();
+        if (line.startsWith(GROUPS_USER)) {
+            foreach (const QString& group, line.mid(strlen(GROUPS_USER) + 1).split(',')) {
+                if (!addUserToGroup(user, group.trimmed())) {
+                    file.close();
+                    return false;
+                }
+            }
+        }
+    }
+    file.close();
+
+    // Droid groups
+    QFile droidDefault("/usr/lib/droid/hw-group.d/default.list");
+    if (droidDefault.open(QIODevice::ReadOnly)) {
+        while (!droidDefault.atEnd()) {
+            line = droidDefault.readLine();
+            if (!line.isEmpty()) {
+                if (!addUserToGroup(user, line.trimmed())) {
+                    droidDefault.close();
+                    return false;
+                }
+            }
+        }
+        droidDefault.close();
+    }
+    QFile droidInet("/usr/lib/droid/hw-group.d/inet.list");
+    if (droidInet.open(QIODevice::ReadOnly)) {
+        while (!droidInet.atEnd()) {
+            line = droidInet.readLine();
+            if (!line.isEmpty()) {
+                if (!addUserToGroup(user, line.trimmed())) {
+                    droidInet.close();
+                    return false;
+                }
+            }
+        }
+        droidInet.close();
+    }
+
+    return true;
+}
+
+bool SailfishUserManager::makeHome(const QString &user)
+{
+    QString destination = QString(USER_HOME).arg(user);
+    QDir source(SKEL_DIR);
+    if (!source.mkdir(destination)) {
+        qWarning() << "User home already exists";
+        return false;
+    }
+
+    struct passwd *pw = getpwnam(user.toLocal8Bit());
+    if (!pw) {
+        qWarning() << "User not found";
+        return false;
+    }
+
+    if (chown(destination.toLocal8Bit(), pw->pw_uid, pw->pw_gid)) {
+        qWarning() << "Home directory ownership change failed";
+        return false;
+    }
+    if (chmod(destination.toLocal8Bit(), HOME_MODE)) {
+        qWarning() << "Home directory permissions change failed";
+        return false;
+    }
+
+    foreach (const QString& file, source.entryList(QDir::Files | QDir::NoDotAndDotDot | QDir::Hidden)) {
```

Comment:
I guess we don't have any use case for copying directories too?

Reply:
I have feeling that there is a ton of subtle gotchas related to this. Would there be some way to use existing code/scripts for skeleton setup?